Added evaluation of logical operators

ggetz opened the pull request:
Added evaluation of logical operators and specs.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -144,6 +158,18 @@ define([
                 node.evaluate = node._evaluateEquals;
             } else if (node._value === '!==') {
                 node.evaluate = node._evaluateNotEquals;
+            } else if (node._value === '<') {
```
Note that all the operators are assuming `number`/`number` comparisons.

* We need to do type checking to make sure, for example, that `4 > 'cat'` is a syntax error.  Please carefully think about our type system and make it reasonably strict, e.g., no unexpected implicit conversions like `true > color('red')`
* We also need to implement operators for `Color` types, e.g., component-wise operators.  If we don't want to do those functions to `Color` (totally cool with me to keep Cesium's API surface area small) then we can just add them a local functions in this file.

ggetz replied:
1. So we should throw a syntax error when we know the types before runtime, but sometimes we need to evaluate the statement to determine the type. Should I throw the same kind of error at runtime for expressions that need to be evaluated?

2. How do we implement color `<`/`>`? Do we do it component wise and give red the highest precedence? This seems arbitrary and something rarely come up. `false < true` isn't really defined, unless you convert booleans to 0/1, so could we have `color('red') < color('green')` be undefined?

pjcozzi replied:
1. Yes, we should throw them at runtime.  When do we need to wait until runtime?  Just for variables?  I suspect we will know everything else at compile time.

2. Those can be syntax errors.  GLSL has `any` and `all` functions, e.g., `any(bvec3(true, false, false)) == true`, but we don't need to introduce all that until we actually need it.  By the way, it is worth a few minutes to check out GLSL, see Pages 3 and 4: https://www.khronos.org/files/webgl/webgl-reference-card-1_0.pdf

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -168,6 +194,38 @@ define([
+
+    Node.prototype._evaluateOr = function(feature) {
+        var left = this._left.evaluate(feature);
+        if (left) {
```
Concisely comment here and below that this implements short-circuiting.

ggetz commented:
I added type checking to the evaluation function. I think figuring out the syntax errors before runtime can be better implemented when we put in the "fast track" evaluation functions, because we'll be doing type checking there.

I wasn't sure of the error message to use when performing an operation that's undefined on certain types, so I used `Operation is undefined`. Is there a way to phrase this to be more informative?

ggetz commented:
Also at this point it might be useful to include the type in the `Node` structure, so we don't have to perform a lot of `typeof`'s and `instanceof`'s.

pjcozzi commented:
> Also at this point it might be useful to include the type in the Node structure, so we don't have to perform a lot of typeof's and instanceof's.

:+1: Want to do it as part of this PR?

ggetz commented:
> Want to do it as part of this PR?

Sure

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -156,6 +182,21 @@ define([
+        var ltype = typeof(left);
+        var rtype = typeof(right);
+        if (ltype !== rtype) {
+            //>>includeStart('debug', pragmas.debug);
```
Throughout the `///` statements can go around the entire `if` block (or function implementation in this case).

ggetz commented:
I added the enum `EvaluationNodeType` and added a `type` property to the `Node` structure.

Currently, since the evaluation functions are more general (ie, they may be a variable or they may have to evaluate child expressions), we cannot only check `node._type`. However this will be useful going forward to assign evaluation functions that are more strongly typed for runtime. 

I also factored out some helper function from `createRuntimeAst` so it's a little more readable.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -170,6 +236,80 @@ define([
+        }
+
+        var right = this._right.evaluate(feature);
+        //>>includeStart('debug', pragmas.debug);
```
This is OK with me (well, hopefully we can catch it at compile time), but there is subtle behavior here.  If the short circuit passes, it would let this error go unnoticed.  I bet JavaScript engines have the same behavior short circuiting is part of the spec.

ggetz replied:
Well, the point of short circuiting is to be more efficient, so wouldn't taking the time to perform that check go against that anyway? Especially if you have to evaluate it to get the type.

pjcozzi replied:
Right, that's why I suggested compile time.  Regardless, I'll make the spec clear that the false expression is not evaluated in this case so errors are not guaranteed to be caught.

pjcozzi commented:
Nice!
